controllers/remote/cluster_cache.go

```diff
@@ -0,0 +1,401 @@
+	}
+
+	cache, err := m.getOrCreateClusterCache(ctx, input.Cluster)
+	if err != nil {
+		return err
+	}
+
+	if err := input.Watcher.Watch(source.NewKindWithCache(input.Kind, cache), input.EventHandler, input.Predicates...); err != nil {
+		return errors.Wrap(err, fmt.Sprintf("error creating watch"))
+	}
+
+	watchesForCluster[wi] = struct{}{}
+
+	return nil
+}
+
+// getOrCreateClusterCache returns the clusterCache for cluster, creating a new ClusterCache if needed.
+func (m *ClusterCacheTracker) getOrCreateClusterCache(ctx context.Context, cluster client.ObjectKey) (*clusterCache, error) {
+	cache := m.getClusterCache(cluster)
+	if cache != nil {
+		return cache, nil
+	}
+
+	return m.newClusterCache(ctx, cluster)
+}
+
+// getClusterCache returns the clusterCache for cluster, or nil if it does not exist.
+func (m *ClusterCacheTracker) getClusterCache(cluster client.ObjectKey) *clusterCache {
+	m.clusterCachesLock.RLock()
+	defer m.clusterCachesLock.RUnlock()
+
+	return m.clusterCaches[cluster]
+}
+
+// newClusterCache creates and starts a new clusterCache for cluster.
+func (m *ClusterCacheTracker) newClusterCache(ctx context.Context, cluster client.ObjectKey) (*clusterCache, error) {
+	m.clusterCachesLock.Lock()
+	defer m.clusterCachesLock.Unlock()
+
+	// If another goroutine created the cache while this one was waiting to acquire the write lock, return that
+	// instead of overwriting it.
+	if c, exists := m.clusterCaches[cluster]; exists {
+		return c, nil
+	}
+
+	config, err := RESTConfig(ctx, m.client, cluster)
+	if err != nil {
+		return nil, errors.Wrap(err, "error fetching REST client config for remote cluster")
+	}
+
+	remoteCache, err := cache.New(config, cache.Options{Scheme: m.scheme})
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating cache for remote cluster")
+	}
+	stop := make(chan struct{})
+
+	cc := &clusterCache{
+		Cache: remoteCache,
+		stop:  stop,
+	}
+	m.clusterCaches[cluster] = cc
+
+	// Start the cache!!!
+	go remoteCache.Start(cc.stop)
+	// Start cluster healthcheck!!!
+	go m.healthCheckCluster(ctx, cluster, config, healthCheckPollInterval, healthCheckRequestTimeout, healthCheckUnhealthyThreshold, "/")
+
+	return cc, nil
+}
+
+func (m *ClusterCacheTracker) deleteClusterCache(cluster client.ObjectKey) {
+	m.clusterCachesLock.Lock()
+	defer m.clusterCachesLock.Unlock()
+
+	delete(m.clusterCaches, cluster)
+}
+
+type ClusterCacheReconciler struct {
+	log     logr.Logger
+	client  client.Client
+	tracker *ClusterCacheTracker
+}
+
+// healthCheckCluster will poll the cluster's API at the path given and, if there are
+// `unhealthyThreshold` consecutive failures, will deem the cluster unhealthy.
+// Once the cluster is deemed unhealthy, the cluster's cache is stopped and removed.
+func (m *ClusterCacheTracker) healthCheckCluster(ctx context.Context, cluster client.ObjectKey, cfg *rest.Config, interval, requestTimeout time.Duration, unhealthyThreshold int, path string) {
```

Comment:
This function has a lot of parameters. Should we consider either an input struct, or a new struct that has members for all these parameters?

Reply:
I've put this into a struct and added a validate function for that struct that populates the optional parts of this struct, does that seem sensible?